NMS-14635: Add alt text to Operation (pt. 4) (#5743)
OpenNMS/opennms: docs/modules/operation/pages/deep-dive/telemetryd/introduction.adoc

The configuration is split in two parts: on one side, listeners and their attached parsers or connectors receive telemetry data that is transported over a specific protocol, and parse the data according to the protocol's specification.
On the other side, adapters pick up the parsed data and enrich it before persisting it.

Queues transport the parsed telemetry data from parsers to adapters by binding a specific parser to a specific adapter.

== Listeners and parsers

Listeners receive sensor data from an external source and hand the data off to parsers, which handle the transport protocol format and extract the transported data.
Listeners do the transport protocol handling, like opening sockets and accepting incoming connections.
The received messages are then passed to the parsers that are configured to the listeners.
They parse the concrete protocol implementation and extract all available information from the messages.
The extracted data is then handed over to the configured queue for further processing.

For example, a listener may open a UDP socket to receive packets, and an assigned parser parses the incoming IPFIX packets.

> Parsers and listeners much work closely together; for that reason, they cannot be combined randomly.

Reviewer: ```suggestion
Parsers and listeners must work closely together; for that reason, they cannot be combined randomly.
```

=== Configuring queues

The following options can help fine-tune queue behavior:

[cols="2,2,3,2"]
|===
| Queue Attribute ({page-component-title})  | Key (Minion/Sentinel)     | Description   | Default Value

| threads
| queue.threads
| Number of threads used for consuming and dispatching messages.
| (2 * _number of cores_)

| queue-size
| queue.size
| Maximum number of messages to keep in memory while waiting to be dispatched.
| 10000

| use-routing-key
| queue.use-routing-key
| Whether to use the routing key when forwarding messages to the broker.
This enforces message ordering.
| true

| batch-size
| batch.size
| Maximum batch size.
When the batch reaches this size, it is dispatched.
| 1000

| batch-interval
| batch.interval
| Maximum batch interval, in milliseconds.
When the batch has been created for longer than this interval, it is dispatched, regardless of the current size.
| 500
|===

TIP: When using Kafka as a message broker, setting `use-routing-key` to `false` balances messages across all partitions.
This can be done safely for flows, but is not supported for metrics when using thresholding (order is required).

> In {page-component-title}, these options can be added to the `<queue>` element as attributes:

Reviewer: ```suggestion
In {page-component-title}, you can add these options to the `<queue>` element as attributes:
```

[source, xml]
----
<queue name="IPFIX" use-routing-key="false">
    ...
</queue>
----

> On a Minion, these options can be added as parser properties, and as adapter properties on Sentinel:

Reviewer: ```suggestion
On a Minion, you can add these options as parser properties, and as adapter properties on Sentinel:
```

parsers.0.queue.use-routing-key=false
----

== Push sensor data through Minion

Listeners and their parsers may run either on {page-component-title} or a Minion, whereas adapters may run either on {page-component-title} or Sentinel.
If a listener and its parsers are running on a Minion, the received messages are dispatched automatically via a queue to associated adapters running in {page-component-title} or Sentinel.
Minions can receive data, but require {page-component-title} or Sentinel to process it.

> .Running Listener on a Minion forwarding packets using the messaging system

Reviewer: ```suggestion
.Running listener on a Minion forwarding packets using the messaging system
```

> image::telemetryd/telemetryd-minion.png["Network overview diagram showing a listener running on a Minion and its related components"]

Reviewer: is it a listener running on minion and the Minion's related components, or a listener and its related components running on a Minion?

Author: I think it's the latter, in which case I'll update the wording.

```suggestion
image::telemetryd/telemetryd-minion.png["Network overview diagram showing a listener running on a Minion, and the listener's related components"]
```
